### ISSUE TITLE
Index use

### ISSUE THREAD

**Issue by vdbergh** [glinscott/fishtest#1242]
I was looking a bit at the indexes in use for Fishest. So I ran the following
```javascript
use fishtest_new
db.runs.find({'finished':true, 'is_green':true}).explain().queryPlanner.winningPlan
```
To my surprise the answer was 
```javascript
{
	"stage" : "COLLSCAN",
	"filter" : {
		"$and" : [
			{
				"finished" : {
					"$eq" : true
				}
			},
			{
				"is_green" : {
					"$eq" : true
				}
			}
		]
	},
	"direction" : "forward"
}
```
The `"stage": "COLLSCAN"` says that mongodb does not use the index `finished_green_runs`. 

Is my local db too small for this? Perhaps mongodb thinks `COLLSCAN` is faster than `IXSCAN`+`FETCH`?

**Comment by vdbergh** [comment]
Some bug reports that might be related: https://jira.mongodb.org/browse/SERVER-26580

**Comment by ppigazzini** [comment]
DEV
```
rs0 [direct: primary] test> use fishtest_new
switched to db fishtest_new
rs0 [direct: primary] fishtest_new> db.runs.find({'finished':true, 'is_green':true}).explain().queryPlanner.winningPlan
{
  stage: 'COLLSCAN',
  filter: {
    '$and': [ { finished: { '$eq': true } }, { is_green: { '$eq': true } } ]
  },
  direction: 'forward'
}
```

**Comment by vdbergh** [comment]
For some reason, if you sort then it seems to use the index
```javascript
fishtest_new> db.runs.find({"finished":true,"is_green":true}).sort({'last_updated':1}).explain().queryPlanner.winningPlan
```
Result:
```javascript
{
  stage: 'FETCH',
  filter: {
    '$and': [ { finished: { '$eq': true } }, { is_green: { '$eq': true } } ]
  },
  inputStage: {
    stage: 'IXSCAN',
    keyPattern: { last_updated: -1, is_green: -1 },
    indexName: 'finished_green_runs',
    isMultiKey: false,
    multiKeyPaths: { last_updated: [], is_green: [] },
    isUnique: false,
    isSparse: false,
    isPartial: true,
    indexVersion: 2,
    direction: 'backward',
    indexBounds: {
      last_updated: [ '[MinKey, MaxKey]' ],
      is_green: [ '[MinKey, MaxKey]' ]
    }
  }
}
```
Now it _does_ use the index, but it appears to not to notice that the filter in the FETCH stage is redundant. The FETCH stage is the same if we use an unrelated index for sorting such as `_id`.

I think this is all a bit scary, since it appears we cannot rely on mongo to just do the right thing.

**Comment by ppigazzini** [comment]
On DEV:
```
rs0 [direct: primary] fishtest_new> db.runs.find({"finished":true,"is_green":true}).sort({'last_updated':1}).explain().queryPlanner.winningPlan
{
  stage: 'FETCH',
  filter: {
    '$and': [ { finished: { '$eq': true } }, { is_green: { '$eq': true } } ]
  },
  inputStage: {
    stage: 'IXSCAN',
    keyPattern: { last_updated: -1, is_green: -1 },
    indexName: 'finished_green_runs',
    isMultiKey: false,
    multiKeyPaths: { last_updated: [], is_green: [] },
    isUnique: false,
    isSparse: false,
    isPartial: true,
    indexVersion: 2,
    direction: 'backward',
    indexBounds: {
      last_updated: [ '[MinKey, MaxKey]' ],
      is_green: [ '[MinKey, MaxKey]' ]
    }
  }
}
```

**Comment by vdbergh** [comment]
After some more experimenting, assuming that the output of `explain()` can be trusted, my conclusion is that partial indexes don't  work and the indexed fields have to be strictly ordered according to the ESR (Equality, Sort, Range) rule. 

With sorting one can sometimes get an index to do something but it is not clear to me what the principle is involved.

**Comment by vdbergh** [comment]
UPDATE: partial indexes do work after all. However it seems the `partialFilterExpression` needs to be part of the fields that are indexed. This is currently not the case for `finished_green_runs`. Currently it looks as follows:
```javascript
  {
    v: 2,
    key: { finished: -1, is_green: -1, last_updated: -1 },
    name: 'finished_green_runs',
    partialFilterExpression: { finished: true, is_green: true }
  }
```
Including `finished` in the indexed fields, and reordering according the ESR rule this becomes
```javascript
  {
    v: 2,
    key: { finished: -1, is_green: -1, last_updated: -1 },
    name: 'finished_green_runs',
    partialFilterExpression: { finished: true, is_green: true }
  }
```
Now suddenly things work
```javascript
use fishtest_new
db.runs.find({"finished":true, "is_green":true}).explain().queryPlanner.winningPlan
```
The output is 
```
{
  stage: 'FETCH',
  inputStage: {
    stage: 'IXSCAN',
    keyPattern: { finished: -1, is_green: -1, last_updated: -1 },
    indexName: 'finished_green_runs',
    isMultiKey: false,
    multiKeyPaths: { finished: [], is_green: [], last_updated: [] },
    isUnique: false,
    isSparse: false,
    isPartial: true,
    indexVersion: 2,
    direction: 'forward',
    indexBounds: {
      finished: [ '[true, true]' ],
      is_green: [ '[true, true]' ],
      last_updated: [ '[MaxKey, MinKey]' ]
    }
  }
}
```
Which is exactly what we expect (no filter in the FETCH stage).

**Comment by vdbergh** [comment]
Solved via #1243
